modules/core/shared/src/main/scala/lucuma/core/geom/OffsetGenerator.scala

```diff
@@ -0,0 +1,138 @@
+ */
+object OffsetGenerator:
+
+  /**
+   * Generates a grid pattern of offsets.
+   */
+  def grid[F[_]: Applicative](nx: PosInt, ny: PosInt, dp: Angle, dq: Angle, center: Offset = Offset.Zero): F[NonEmptyList[Offset]] =
+    val pStart = -(nx.value - 1) / 2.0
+    val qStart = (ny.value - 1) / 2.0
+
+    val offsets =
+      for
+        j <- 0 until ny.value
+        i <- 0 until nx.value
+      yield
+        val pOffset = dp * (pStart + i)
+        val qOffset = dq * (qStart - j)
+        center + Offset(pOffset.p, qOffset.q)
+
+    // can't be empty as nx and ny are 1 or more
+    NonEmptyList.fromListUnsafe(offsets.toList).pure[F]
+
+  /**
+   * Generates a random pattern of offsets.
+   * Creates a grid and randomly places points near each grid element.
+   */
+  def random[F[_]: {Monad, Random as R}](num: PosInt, size: Angle, center: Offset = Offset.Zero): F[NonEmptyList[Offset]] =
+    val n = PosInt.unsafeFrom(ceil(sqrt(num.value.toDouble)).toInt)
+    val d = size * (1.0 / n.value.toDouble)
+
+    def randomizedPoints(grid: NonEmptyList[Offset]) =
+      grid.traverse: point =>
+        for
+          pJitter <- R.nextDouble.map(x => d * (x - 0.5))
+          qJitter <- R.nextDouble.map(x => d * (x - 0.5))
+        yield point + Offset(pJitter.p, qJitter.q)
+
+    for
+      gridPoints       <- grid[F](n, n, d, d, center)
+      randomizedPoints <- randomizedPoints(gridPoints)
+      shuffled         <- R.shuffleList(randomizedPoints.toList)
+    yield NonEmptyList.fromListUnsafe(shuffled.take(num.value))
+
+  /**
+   * Generates a Fermat spiral pattern of offsets.
+   * Uses the golden angle for optimal packing.
+   */
+  def spiral[F[_]: {Monad, Random as R}](num: PosInt, size: Angle, center: Offset = Offset.Zero): F[NonEmptyList[Offset]] =
+    val θ = Angle.fromDoubleDegrees(137.50776)  // golden angle
+
+    val offsets = R.nextDouble.map(x => Angle.fromDoubleRadians(x * 2.0 * Pi)).map: φ =>
+      // Calculate scale factor
+      val arcsecs = Angle.signedDecimalArcseconds.get(size).toDouble
+      val θʹ = θ * num.value.toDouble
+      val α = arcsecs / 2.0 / sqrt(num.value * (θʹ.cos * θʹ.cos + θʹ.sin * θʹ.sin))
+
+      (0 until num.value).map: n =>
+        val sqrtN = sqrt(n.toDouble)
+        val angle = θ * n.toDouble + φ
+        val p = α * sqrtN * angle.cos
+        val q = α * sqrtN * angle.sin
+
+        center + Offset(Angle.fromDoubleArcseconds(p).p, Angle.fromDoubleArcseconds(q).q)
+
+    offsets.map: o =>
+      NonEmptyList.fromListUnsafe(o.toList)
+
+  /**
+   * Generates a 1D grid pattern of P components.
+   */
+  def gridP[F[_]: Applicative](num: PosInt, step: Angle, center: Offset.P = Offset.P.Zero): F[NonEmptyList[Offset.P]] =
+    val start = -(num.value - 1) / 2.0
+
+    val components =
+      (0 until num.value).map: i =>
+        val pOffset = step * (start + i)
+        center + Offset.P(pOffset)
+
+    NonEmptyList.fromListUnsafe(components.toList).pure[F]
```

Comment:
This one also could maybe just return a `NonEmptyList` directly?  Or maybe `grid` could be done in terms of `gridP` and `gridQ` or vice versa.  Like `gridP(nx, dp, center).zipWith(gridQ(ny, dq, center)((p, q) => Offset(p, q))` or something.  The same for `random` / `randomP` / `randomQ` perhaps.